gliner/trainer.py

```diff
@@ -0,0 +1,256 @@
+            
+        self.register_buffer('step', torch.tensor(0))
+
+        self.checkpoint_every_step = checkpoint_every_step 
+        self.checkpoint_every_epoch = checkpoint_every_epoch 
+        self.checkpoint_folder = Path(checkpoint_folder)
+        self.checkpoint_folder.mkdir(exist_ok = True, parents = True)
+
+    @contextmanager
+    @beartype
+    def trackers(
+        self,
+        project_name: str,
+        run_name: Optional[str] = None,
+        hps: Optional[dict] = None
+    ):
+        assert self.use_wandb_tracking
+
+        self.accelerator.init_trackers(project_name, config = hps)
+
+        if exists(run_name):
+            self.accelerator.trackers[0].run.name = run_name
+
+        yield
+        self.accelerator.end_training()
+
+    def log(self, **data_kwargs):
+        self.accelerator.log(data_kwargs, step = self.step.item())
+
+    @property
+    def device(self):
+        return self.unwrapped_model.device
+
+    @property
+    def is_main(self):
+        return self.accelerator.is_main_process
+
+    @property
+    def unwrapped_model(self):
+        return self.accelerator.unwrap_model(self.model)
+
+    @property
+    def is_local_main(self):
+        return self.accelerator.is_local_main_process
+
+    def wait(self):
+        return self.accelerator.wait_for_everyone()
+
+    def print(self, msg):
+        return self.accelerator.print(msg)
+
+    def save(self, path, overwrite = True):
+        path = Path(path)
+        assert overwrite or not path.exists() 
+        pkg = dict(
+            model = self.unwrapped_model.state_dict(),
+            optimizer = self.optimizer.state_dict(),
+            step = self.step.item() 
+        ) 
+        torch.save(pkg, str(path))
+
+    def load(self, path):
+        path = Path(path)
+        assert path.exists()
+
+        pkg = torch.load(str(path)) 
+        self.model.load_state_dict(pkg['model'])
+        self.optimizer.load_state_dict(pkg['optimizer'])
+        self.step.copy_(pkg['step'])
+  
+    def train(self, 
+              num_epochs = None, 
+              num_steps = None,
+              num_warmup_steps: int = 1000):
+        assert not (num_epochs is None and num_steps is None) , 'you must specify either num_epochs or num_steps'  
+        step = self.step.item()
+        
+        total_steps = num_epochs * len(self.train_loader) if num_steps is None else num_steps
+        num_epochs = default(num_epochs, total_steps // len(self.train_loader))
+        
+        total_steps += default(step, 0)
+         
+        self.print(f"Training for {total_steps} steps which is {num_epochs} epochs.") 
+        epoch_size = len(self.train_loader)
+        self.model.train()  
+        
+        self.scheduler = get_cosine_schedule_with_warmup(
+            self.optimizer,
+            num_warmup_steps=num_warmup_steps,
+            num_training_steps=total_steps
+        )
+        
+        for epoch in range(num_epochs): 
```

Comment:
Adding `.zero_grad()` at the start of the training loop prevent some bug for me.